Several updates/changes

tschiemer opened the pull request:
Hi! Nice library, and thanks!

As we ran into some problems and also needed some customizations we made a few changes which you'll hopefully like..
1. When run out of the box, the php imap functions can raise some errors because the communication with the server is considered unsafe. These messages would originally be flushed after any script is run and might thus interfere with any system output. We wrapped this, to catch any errors/alerts and store them into Downloader::$errors, Downloader::$alerts respectively.

example imap error: SECURITY PROBLEM: insecure server advertised AUTH=PLAIN
1. Option to automatically create the processed folder, if it does not exist yet.
2. Option to set the inbox folder, in case not default.
3. Extended / customized the retrieved data of emails, such that either 1. no actual data is retrieved, 2. the overview (imap_fetchoverview) is retrieved 3. the body (imap_fetchbody) is retrieved 4. the complete headers (imap_fetchheader) is retrieved. Any of these can be arbitrarily combined.
4. Introduced more complicated processing actions whereof any can be set as default processing method  or can be overriden on a per-email basis:1. move message to arbitrary folder 2. delete message 3. call a custom callback that is passed the mailbox resource and emailIndex.

The default behaviour was respected so updating should not require any code changes.

Commit:
Added php alerts/errors suppression

Commit:
Inbox folder is customizable

Commit:
1. extended email parts
2. define email parts to actually fetch (overview, body, headers)

Commit:
automake of processed folder

Commit:
customizable process action

Commit:
updated readme

Commit:
updated readme

Commit:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	Readme.md

tschiemer commented:
Ok, so looks like coding style violations. I'll have that fixed..

Commit:
coding style updates

Commit:
coding style updates

Commit:
backwards php compatibility in default arguments

tomaj commented on Readme.md:
````diff
@@ -45,4 +45,66 @@ $downloader->fetch($criteria, function(Email $email) {
+``` php
+
+// define what you consider your inbox folder..
+$downloader->setInboxFolder('INDBOX');
````
Is this typo? or is it purpose to set it to "INDBOX" ?

tomaj commented on Readme.md:
```diff
@@ -45,4 +45,66 @@ $downloader->fetch($criteria, function(Email $email) {
+//  FETCH_HEADERS       complete email headers as string
+// default = FETCH_OVERVIEW | FETCH_BODY
+
+$fetchParts = Downloader::FETCH_OVERVIEW | Downloader::FETCH_BODY | Downloader::FETCH_HEADERS;
```
This variable is never used in this example. I understand that you will be able to set behaviour for downloader based on combination of this settings. But I think it will be valuable to use it in example, not just define variable.

tomaj commented on src/Downloader.php:
```diff
@@ -24,46 +54,176 @@ public function __construct($host, $port, $username, $password)
+        return $this->errors;
+    }
+
+    public function fetch(MailCriteria $criteria, $callback, $fetchParts = null)
```
This method starts looking like very long and complicated method. This should be refactored. It is not part of this merge request ;-) i am just writing it for future myself.

tomaj commented:
Thank you for you participation. This pull request looks great and this should advance this library to next level. Thank. 

Can you update your code to PSR-2 (https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide.md). Source files and also tests.

Hope we will be able to merge it soon.

Commit:
- PSR2 compatibility
- PHP5.3 backwards compatibility (no finally statements..)
- documentation
- minor code changes

Commit:
phpunit test php53 compatible

Commit:
phpunit test php53 compatible +1

Commit:
phpunit test php53 compatible +2

Commit:
php53 compatibility

Commit:
php53 compatibility

Commit:
- Downloader::fetch() phpdoc
- Downloader::checkProcessedFolder(): throw exception on autocreate error

Commit:
- refactoring of Downloader::fetch()
- consistent differentiation between Exception and ImapException

Commit:
namespace bugfix

tschiemer commented:
Hi! Thanks for your comments and hints.

It was more of an effort than expected, in particular the backwards compatibility.. meow

I yet made some changes:
- extended+corrected readme
- refactored `fetch()` slightly to make it more readable
- made the exception throwing a bit more complete and consistent to differentiate between application exceptions and imap exceptions

I would think it should be fine now :)

tschiemer commented:
err no, wait. there might be a bug in the automatic creation of folders..

Commit:
bugfix checking of folders and autocreation

tschiemer commented:
voilà
